File: platforms/ios/devkit/Sources/NativeblocksDevkit/DevKit.swift
import Foundation
import NativeblocksRuntime
import UserNotifications

/// `DevKit` provides live features such as real-time updates, remote logging and screen
/// management for Nativeblocks applications.
/// It is specifically designed to work with the Nativeblocks Studio environment.
///
/// Example:
/// ```swift
/// let devKit = DevKit.Builder()
///     .keepScreenOn()
///     .autoConnect()
///     .logTracking()
///     .build()
///
/// NativeblocksManager.getInstance(name: instanceName).provideKit(devKit)
/// ```
public final class DevKit: Kit {

    private static var configuredInstanceName: String?

    // MARK: - Properties
    /// Keeps the screen active and prevents it from locking.
    private let keepScreenOn: Bool

    /// Automatically connects to the environment during setup.
    private let autoConnect: Bool

    /// Enables remote logging to Nativeblocks Studio.
    private let logTracking: Bool

    /// Reference to the dev logger instance
    private var devLogger: DevLogger?

    // MARK: - Initialization

    /// Private initializer to enforce Builder pattern usage.
    /// - Parameters:
    ///   - keepScreenOn: A flag to keep the screen active (default: `false`).
    ///   - autoConnect: A flag to enable automatic connection (default: `false`).
    ///   - logTracking: A flag to enable remote logging (default: `false`).
    private init(keepScreenOn: Bool, autoConnect: Bool, logTracking: Bool) {
        self.keepScreenOn = keepScreenOn
        self.autoConnect = autoConnect
        self.logTracking = logTracking
    }

    /// Attaches `DevKit` to the given instance based on the provided `NativeblocksEdition`.
    ///
    /// - Parameters:
    ///   - instanceName: The unique identifier for this DevKit instance.
    ///   - edition: The `NativeblocksEdition` to configure the environment.
    ///
    /// - Warning: This function should not be called directly by developers. It is invoked automatically
    /// during the initialization process of Nativeblocks.
    public func attach(instanceName: String, edition: NativeblocksEdition) {
        if let configured = Self.configuredInstanceName, configured != instanceName {
            fatalError("DevKit is already set up with '\(configured)'. Cannot reconfigure with '\(instanceName)'.")
        }
        if Self.configuredInstanceName == nil {
            Self.configuredInstanceName = instanceName
        }

        DevKitLogger.debug("DevKit", "attach")
        let environment: DevKitEnvironment

        switch edition {
        case .cloud(_, let apiKey, let developmentMode):
            if !developmentMode {
                fatalError("DevKit doesn't support production mode, please disable it or use on the developmentMode")
            }
            environment = DevKitEnvironment(
                instanceName: instanceName,
                apiKey: apiKey,
                screenSharing: false,
                autoConnect: autoConnect
            )
        default:
            fatalError("DevKit is supported only for cloud edition")
        }

        if keepScreenOn {
            LockScreen.lockScreenOff()
        }

        DevKitInjector.initialize(environment: environment)

        if logTracking {
            devLogger = LoggingModule.provideDevLogger()
            if let logger = devLogger {
                _ = NativeblocksManager.getInstance(name: instanceName)
                    .provideEventLogger(loggerType: "DevKit:\(instanceName)", logger: logger)
            }
        }

        LiveService.start()
    }

    /// Manually authorizes a user with the DevKit module.
    ///
    /// This method is used to authenticate a user with a backend endpoint and a valid token and real-time connection endpoint.
    /// It should be called before using any DevKit features that require authentication.
    /// - Parameters:
    ///   - endpoint: The base URL of your backend API.
    ///   - token: The user authentication token (e.g., JWT).
    ///   - realtimeEndpoint: The real-time server endpoint used for live features (e.g., logs, screen sharing).
    public func userAuthorization(endpoint: String, token: String, realtimeEndpoint: String) {
        DevKitInjector.userAuthorization(endpoint: endpoint, token: token, realtimeEndpoint: realtimeEndpoint)
    }

    /// Manual user logout
    public func userLogout() {
        DevKitInjector.userLogout()
    }

    /// Detaches the `DevKit` instance and cleans up resources.
    /// - Note: If `keepScreenOn` is enabled, the screen lock will be re-enabled.
    public func detach(instanceName: String) {
        DevKitLogger.debug("DevKit", "detach")
        if keepScreenOn {
            LockScreen.lockScreenOn()
        }
        devLogger = nil
        DevKitInjector.destroy()
        Self.configuredInstanceName = nil
    }

    // MARK: - Builder

    /// Builder class for constructing `DevKit` instances with a fluent API.
    ///
    /// Example:
    /// ```swift
    /// let devKit = DevKit.Builder()
    ///     .keepScreenOn()
    ///     .autoConnect()
    ///     .logTracking()
    ///     .build()
    /// ```
    public class Builder {
        private var _keepScreenOn: Bool = false
        private var _autoConnect: Bool = false
        private var _logTracking: Bool = false

        public init() {}

        /// Enables the keep screen on feature.
        /// When enabled, the device screen will stay on and won't auto-lock.
        /// - Returns: The builder instance for method chaining.
        @discardableResult
        public func keepScreenOn() -> Builder {
            self._keepScreenOn = true
            return self
        }

        /// Enables automatic connection to the development environment.
        /// When enabled, the SDK will automatically connect to Nativeblocks Studio on startup.
        /// - Returns: The builder instance for method chaining.
        @discardableResult
        public func autoConnect() -> Builder {
            self._autoConnect = true
            return self
        }

        /// Enables remote logging to Nativeblocks Studio.
        /// When enabled, application logs will be sent to Studio for debugging.
        /// - Returns: The builder instance for method chaining.
        @discardableResult
        public func logTracking() -> Builder {
            self._logTracking = true
            return self
        }

        /// Builds and returns the configured `DevKit` instance.
        /// - Returns: A new `DevKit` instance with the configured options.
        public func build() -> DevKit {
            return DevKit(
                keepScreenOn: _keepScreenOn,
                autoConnect: _autoConnect,
                logTracking: _logTracking
            )
        }
    }
}

File: platforms/android/devkit/devkit/src/main/java/io/nativeblocks/devkit/DevKit.kt
package io.nativeblocks.devkit

import android.content.Context
import androidx.activity.ComponentActivity
import io.nativeblocks.runtime.api.NativeblocksEdition
import io.nativeblocks.runtime.api.NativeblocksManager
import io.nativeblocks.runtime.api.provider.kit.Kit
import io.nativeblocks.devkit.feature.live.presenter.LiveService
import io.nativeblocks.devkit.feature.logging.DevLogger
import io.nativeblocks.devkit.lib.permission.NotificationPermissionContractor
import io.nativeblocks.devkit.lib.permission.KeepScreenOnContractor
import io.nativeblocks.devkit.session.DevKitEnvironment
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job

/**
 * Devkit provides live features such as real-time updates, screen sharing,
 * notification permissions, and screen management for Nativeblocks applications.
 *
 * Use [DevKit.Builder] to create and configure an instance.
 *
 * Example:
 * ```
 * // In your Activity
 * val devkit = DevKit.Builder(this)
 *     .keepScreenOn()
 *     .autoConnect()
 *     .logTracking()
 *     .build()
 *
 * NativeblocksManager.getInstance(instanceName).provideKit(devkit)
 * ```
 *
 * DevKit automatically handles notification permission requests using the provided Activity.
 * When permission is required, DevKit will launch the system permission dialog.
 */
class DevKit private constructor(
    private val activity: ComponentActivity,
    private val keepScreenOn: Boolean,
    private val autoConnect: Boolean,
    private val logTracking: Boolean,
) : Kit {

    companion object {
        private var configuredInstanceName: String? = null
    }

    private var job = Job()
    private var coroutineScope = CoroutineScope(Dispatchers.IO + job)
    private val keepScreenOnContractor = KeepScreenOnContractor()
    private val notificationPermission = NotificationPermissionContractor(activity)
    private var devLogger: DevLogger? = null

    /**
     * Sets up Devkit with the given [context] and [edition].
     *
     * @param context The Android [android.content.Context] used to initialize the live service.
     * @param edition The [NativeblocksEdition] to configure the environment.
     * @param instanceName The instance name for this Devkit configuration.
     *
     * @throws IllegalStateException if Devkit is already set up with a different instance.
     * @throws IllegalArgumentException if production mode is used (only development mode is supported).
     * @throws IllegalAccessException if the Community edition is used (only Cloud edition is supported).
     */
    override fun attach(context: Context, instanceName: String, edition: NativeblocksEdition) {
        synchronized(this) {
            if (configuredInstanceName != null && configuredInstanceName != instanceName) {
                throw IllegalStateException("DevKit is already set up with '$configuredInstanceName'. Cannot reconfigure with '$instanceName'.")
            }
            if (configuredInstanceName == null) {
                configuredInstanceName = instanceName
            }
        }

        job = Job()
        coroutineScope = CoroutineScope(Dispatchers.IO + job)

        val environment = when (edition) {
            is NativeblocksEdition.Cloud -> {
                if (!edition.developmentMode) {
                    throw IllegalArgumentException("Devkit doesn't support production mode, please disable it or use development mode")
                }
                DevKitEnvironment(
                    instanceName = instanceName,
                    apiKey = edition.apiKey,
                    screenSharing = false,
                    autoConnect = autoConnect,
                )
            }

            is NativeblocksEdition.Community -> {
                throw IllegalAccessException("Devkit is supported only for cloud edition")
            }
        }

        DevKitInjector.init(context, environment)

        if (logTracking) {
            devLogger = DevKitInjector.get().koin.get<DevLogger>()
            devLogger?.let {
                NativeblocksManager.getInstance(instanceName).provideEventLogger("DevKit:$instanceName", it)
            }
        }

        if (notificationPermission.isPermissionRequired()) {
            notificationPermission.requestPermission { isGranted ->
                if (isGranted) {
                    LiveService.start(context)
                }
            }
        } else {
            LiveService.start(context)
        }

        if (keepScreenOn) {
            NativeblocksManager.getInstance(instanceName).provideActionContractor(keepScreenOnContractor)
        }
    }

    /**
     * Authenticates the user manually by providing essential credentials and endpoints required
     * for establishing a secure and authenticated connection to the Nativeblocks cloud service.
     *
     * @param endpoint The base URL for your Nativeblocks backend.
     * @param token A valid user authentication token issued by the backend.
     * @param realtimeEndpoint The WebSocket endpoint for real-time interactions such as live updates or signaling.
     *
     * @throws IllegalArgumentException If any of the parameters are empty or malformed.
     */
    suspend fun userAuthorization(endpoint: String, token: String, realtimeEndpoint: String) {
        DevKitInjector.userAuthorization(endpoint, token, realtimeEndpoint)
    }

    /**
     * Logs out the currently authorized user manually by clearing all sensitive tokens
     * and connection data from memory.
     */
    suspend fun userLogout() {
        DevKitInjector.userLogout()
    }

    /**
     * Cleans up resources and cancels any active operations associated with Devkit.
     */
    override fun detach(instanceName: String) {
        job.cancel()
        devLogger = null
        notificationPermission.unregister()
        DevKitInjector.destroy()
        configuredInstanceName = null
    }

    /**
     * Builder class for creating [DevKit] instances with a fluent API.
     *
     * @param activity The [ComponentActivity] used for permission requests.
     *
     * Example:
     * ```
     * val devkit = DevKit.Builder(this)
     *     .keepScreenOn()
     *     .autoConnect()
     *     .build()
     * ```
     */
    class Builder(
        private val activity: ComponentActivity
    ) {
        private var keepScreenOn: Boolean = false
        private var autoConnect: Boolean = false
        private var logTracking: Boolean = false

        /**
         * Sets whether to keep the device screen on during live sessions.
         *
         * @return This builder instance for method chaining.
         */
        fun keepScreenOn() = apply {
            this.keepScreenOn = true
        }

        /**
         * Sets whether to automatically connect to the live service.
         *
         * @return This builder instance for method chaining.
         */
        fun autoConnect() = apply {
            this.autoConnect = true
        }

        /**
         * Enables logging to transmit log events to Nativeblocks Studio.
         *
         * @return This builder instance for method chaining.
         */
        fun logTracking() = apply {
            this.logTracking = true
        }

        /**
         * Builds and returns a [DevKit] instance with the configured settings.
         *
         * @return A new [DevKit] instance.
         */
        fun build(): DevKit {
            return DevKit(
                activity = activity,
                keepScreenOn = keepScreenOn,
                autoConnect = autoConnect,
                logTracking = logTracking,
            )
        }
    }
}